torch/csrc/utils/tensor_numpy.cpp

```diff
@@ -347,6 +354,198 @@ at::Tensor tensor_from_cuda_array_interface(PyObject* obj) {
+std::string get_buffer_str(void* buffer, int64_t start, int64_t end){
+  T* array = static_cast<T*>(buffer); 
+  std::string output = "(";
+  for (int i=start; i<end; ++i) {
+    output += (std::to_string(array[i]) + ", ");
+  }
+  return output + ")";
+}  
+
+
+bool _extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, size_t (&numels)[], size_t shape_idx, std::vector<PyArrayObject*> & array_ptr_storage) {
+  if (PyArray_Check(obj)) {
+    if (numels[shape_idx] != PyArray_SIZE((PyArrayObject*) obj)) {
+      auto true_shape_str = get_buffer_str<npy_intp> (static_cast<void*>(PyArray_DIMS((PyArrayObject*) obj)), 0, PyArray_NDIM((PyArrayObject*) obj)) + std::string(")");
+      auto expected_shape_str = get_vector_str(shape, shape_idx);
+      auto err_str = std::string("expected numpy array of shape ") + expected_shape_str + std::string(" at dim ") + std::to_string(shape_idx) + std::string("(got ") + true_shape_str;
+      throw ValueError(err_str);  
+    }
+    array_ptr_storage.push_back((PyArrayObject*) obj);
+    return true;
+  } 
+  else if (PySequence_Check(obj)) {
+    auto seq = THPObjectPtr(PySequence_Fast(obj, "not a sequence"));
+    if (!seq) throw python_error();
+    auto seq_len = PySequence_Fast_GET_SIZE(seq.get());
+    if (seq_len != shape[shape_idx]) {
+      throw ValueError("expected sequence of length %lld at dim %lld (got %lld)",
+        (long long)shape[shape_idx], (long long)shape_idx, (long long)seq_len);
+    }
+    PyObject** items = PySequence_Fast_ITEMS(seq.get());
+    for (size_t i=0; i<seq_len; ++i) {
+      if (!_extract_ndarrays(items[i], shape, numels, shape_idx+1, array_ptr_storage)) {
+        return false;
+      }
+    }
+  } 
+  else {
+    return false;
+  }
+  return true;
+}
+
+
+bool extract_ndarrays(PyObject* obj, std::vector<int64_t> & shape, std::vector<PyArrayObject*> & array_ptr_storage) {
+
+  auto ndims = shape.size();
+  size_t numels[ndims];
+  numels[ndims-1] = shape[ndims-1];
+
+  for (int i=ndims-2; i>=0; --i) {
+    numels[i] = shape[i] * numels[i+1];
+  }
+
+  return _extract_ndarrays(obj, shape, numels, 0, array_ptr_storage);
+}
+
+
+template <typename T>
+void memcpy_cpu(void* dst, const void* src, size_t num_items, bool isaligned, int itemsize=1) {
+  if (isaligned) {
+    at::parallel_for(
+      0,
+      num_items,
+      at::internal::GRAIN_SIZE,
+      [=](size_t begin, size_t end) {  
+        T* src_tmp = ((T*) src) + begin;
+        T* dst_tmp = ((T*) dst) + begin;
+        auto last = src_tmp+(end-begin);
+
+        while(src_tmp < last){
+          *dst_tmp = *src_tmp;
+          ++src_tmp;
+          ++dst_tmp;
+        }
+      });
+  } 
+  else {
+    at::parallel_for(
+      0,
+      num_items*itemsize,
+      at::internal::GRAIN_SIZE,
+      [=](size_t begin, size_t end) {  
+        char* src_tmp = ((char*) src) + begin;
+        char* dst_tmp = ((char*) dst) + begin;
+        auto last = src_tmp+(end-begin);
+
+        while(src_tmp < last){
```

Comment:
How would this handle discontiguous arrays? Like an array with an innermost stride that isn't the bitwidth of its dtype? (Put another way, what about a view of an array with every other element missing?)